### PR TITLE
Improve x509 module

### LINKED ISSUE #34117: x509 module overwrites private key in bundled certificates
### x509 module overwrites private key in bundled certificates

The x509 module does not support having the private key, DH params and certificate in the
same file, the `write_pem` function does not check the existing file it simply
clobbers the file with the new content.

To should be possible to emulate the following openssl command using the x509
module currently it is not.

`openssl req -x509 -nodes -days 3650 -sha256 -subj '/CN=baruwa.com' -newkey rsa:4096 -keyout /etc/pki/mycert.pem -out /etc/pki/mycert.pem`

It should also be possible to have DH params within the same single file.

Software such as stunnel use this kind of format and there is no option to
specifiy seperate private key and dh param files.

### Setup
Not relevant

### Steps to Reproduce Issue
`salt '*' x509.create_private_key path=/etc/pki/mycert.pem`
`salt '*' x509.create_csr path=/etc/pki/mycert.csr public_key='/etc/pki/mycert.pem' CN='baruwa.com'`
`salt '*' x509.create_certificate path=/etc/pki/mycert.pem signing_private_key='/etc/pki/mycert.pem' csr='/etc/pki/mycert.csr'`

### Versions Report
Not relevant

### PR DESCRIPTION
**Pull request by akissa**
## What does this PR do?

Adds the ability to create and store private keys, certificates in the same file.
If DH params are found in the file they will be preserved.

It also cleans up the module (Pylint).

Fixes a rare exception that is generated as signing_cert is an optional value.
## What issues does this PR fix or reference?

This PR fixes #34117

### CODE REVIEW

**Comment by cachedout** [comment]
I don't think we want Salt to depend on OpenSSL being installed. Can we move the OpenSSL import into a try/except and then just return an error from that function if it's not available? Of course, it could also be up in __virtual__ (which I personally would prefer) but from your comment in the code, it looks like you have it down in this function because it's the only one which requires it. Either way, this will blow up with ImportErrors unless we catch the exception on a failed import.

**Comment by akissa** [comment]
@cachedout I am not sure which line number you are referring to, my changes do not introduce any new imports

**Comment by cachedout** [comment]
Hi @akissa 

Apologies for the confusion. I'm talking about this line: https://github.com/saltstack/salt/pull/34119/files#diff-4dc56ad912db8d1856c286cd18fc6cdeR850

Clearly, you didn't change this directly but some change here is now exercising that function in the test suite. It still needs to be fixed before we can get this in though, I'm afraid. We can't do a direct import like this of something of a library that we don't dep on.

**Comment by akissa** [comment]
The dep on openssl has been there from where the history of this module starts, https://github.com/saltstack/salt/blob/c53b31342393d1da08ef28d24a476baceb495080/salt/modules/x509.py#L664

Could you indicate what the test suite issue is and how this should be fixed.

**Comment by akissa** [comment]
The dep was introduced in https://github.com/saltstack/salt/commit/7252c708940e080e2cc7b318868debf4fb959d7c

**Comment by cachedout** [comment]
@akissa Thank you for tracking that down.

@clinta Just a ping to bring you into this conversation, since this is your commit.

**Comment by akissa** [comment]
I have commited https://github.com/akissa/salt/commit/d3849489669062ca2cb7c85f7b0e2fc8890522cb which should address the openssl import and binary command concerns

**Comment by cachedout** [comment]
@akissa Thanks very much for cleaning that up even though it wasn't caused by you directly. Much appreciated.

File: salt/modules/x509.py [hunk 1]
```diff
@@ -680,7 +697,24 @@ def write_pem(text, path, pem_type=None):
...
+        try:
+            _dhparams = get_pem_entry(_filecontents, 'DH PARAMETERS')
+        except salt.exceptions.SaltInvocationError:
+            pass
+        try:
+            _private_key = get_pem_entry(_filecontents, 'RSA PRIVATE KEY')
+        except salt.exceptions.SaltInvocationError:
+            pass
+    with salt.utils.fopen(path, 'w') as _fp:
+        if pem_type and pem_type == 'CERTIFICATE' and _private_key:
```
Comment (clinta) [inline salt/modules/x509.py]: This is going to write out the private key even if the requested type to write out was CERTIFICATE? I don't think that's a safe assumption to make.

Perhaps pem_type could take multiple values so you could pass in both CERTIFICATE and RSA PRIVATE KEY, but if only certificate is specified, I think only certificate should be written.

**Comment by akissa** [comment]
@clinta No, it is going to preserve the private key that already exists in the file.703 reads up the existing file, the next lines extract the private key and dhparams from the file. The certificate is then sandwitched between the two.

**Comment by akissa** [comment]
@clinta If there is no existing private key in the file, the normal existing behavior is exhibited.

**Comment by akissa** [comment]
This is basically what i am trying to achieve, current logic causes the `stunnel-sign-cert` state to overwrite the key created in `stunnel-create-host-key`.

```
stunnel-create-host-key:
  x509.private_key_managed:
    - name: /etc/pki/baruwa/certs/baruwa.com.pem
    - bits: 4096
    - verbose: False
stunnel-create-csr:
  x509.csr_managed:
    - name: /etc/pki/baruwa/csr/baruwa.com.csr
    - public_key: /etc/pki/baruwa/certs/baruwa.com.pem
    - CN: 192.168.1.1
    - require:
      - x509: stunnel-create-host-key
stunnel-sign-cert:
  x509.certificate_managed:
    - name: /etc/pki/baruwa/certs/baruwa.com.pem
    - signing_private_key: /etc/pki/baruwa/certs/baruwa.com.pem
    - csr: /etc/pki/baruwa/csr/baruwa.com.csr
    - days_valid: 3650
    - require:
      - x509: stunnel-create-csr
```

**Comment by clinta** [comment]
Oh, I see, this is in effect avoiding overwriting the entire file every time.

I've done similar in my own states, but I did it by adding an additional state to accumulate the separate pem files into one larger file. So I can see how this is useful.

I submitted a PR to your branch to gate this behavior behind a new overwrite option which can be passed to the state. I'd prefer that to changing the currently expected behavior where certificate_managed manages the entire file.

**Comment by akissa** [comment]
@clinta Merged

**Comment by akissa** [comment]
@cachedout The changes have been made, could you please review.

**Comment by clinta** [comment]
Looks good to me

**Comment by akissa** [comment]
@cachedout Thanks

**Comment by akissa** [comment]
@clinta Thanks

**Pull request merged**